Backported from https://github.com/xiph/opus/commit/51b99f6d3c6f455ec32cb9c323ccc0ecd9d4e8c5
Add sse4.1 flag when using clang-cl on MSVC targets

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 8535ad5f..702a7ec8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -14,6 +14,14 @@ include(GNUInstallDirs)
 include(CMakeDependentOption)
 include(FeatureSummary)
 
+# Detect clang-cl
+set(MSVC_CLANG_CL false BOOL)
+if (MSVC)
+    if("${CMAKE_C_COMPILER_ID}" MATCHES "Clang")
+        set(MSVC_CLANG_CL true BOOL)
+    endif()
+endif()
+
 set(OPUS_BUILD_SHARED_LIBRARY_HELP_STR "build shared library.")
 option(OPUS_BUILD_SHARED_LIBRARY ${OPUS_BUILD_SHARED_LIBRARY_HELP_STR} OFF)
 if(OPUS_BUILD_SHARED_LIBRARY OR BUILD_SHARED_LIBS OR OPUS_BUILD_FRAMEWORK)
@@ -484,20 +492,20 @@ if(NOT OPUS_DISABLE_INTRINSICS)
         add_sources_group(opus lpcnet ${dnn_sources_sse4_1})
       endif()
       target_compile_definitions(opus PRIVATE OPUS_X86_MAY_HAVE_SSE4_1)
-      if(NOT MSVC)
+      if(NOT MSVC OR MSVC_CLANG_CL)
         set_source_files_properties(${celt_sources_sse4_1} ${silk_sources_sse4_1} ${dnn_sources_sse4_1} PROPERTIES COMPILE_FLAGS -msse4.1)
       endif()
 
       if(OPUS_FIXED_POINT)
         add_sources_group(opus silk ${silk_sources_fixed_sse4_1})
-        if(NOT MSVC)
+        if(NOT MSVC OR MSVC_CLANG_CL)
           set_source_files_properties(${silk_sources_fixed_sse4_1} PROPERTIES COMPILE_FLAGS -msse4.1)
         endif()
       endif()
     endif()
     if(OPUS_X86_PRESUME_SSE4_1)
       target_compile_definitions(opus PRIVATE OPUS_X86_PRESUME_SSE4_1)
-      if(NOT MSVC)
+      if(NOT MSVC OR MSVC_CLANG_CL)
         target_compile_options(opus PRIVATE -msse4.1)
       endif()
     endif()
